File: extra-code/stage2_code/cpp_hardware_tutorial/src/rectangle_area_server.cpp
// ============================================================
// 自定义 Service 示例:矩形面积计算服务端
// 演示:使用自己定义的 hardware_interfaces/srv/ComputeRectangleArea 接口
// (Request: length、width;Response: area)提供 /compute_rectangle_area 服务
// 运行:ros2 run cpp_hardware_tutorial rectangle_area_server
// 测试(命令一行输入):ros2 service call /compute_rectangle_area
//       hardware_interfaces/srv/ComputeRectangleArea "{length: 3.0, width: 4.0}"
// ============================================================

#include <memory>

// 自定义 Service 的头文件同样由 rosidl 构建接口包时自动生成
#include "hardware_interfaces/srv/compute_rectangle_area.hpp"
#include "rclcpp/rclcpp.hpp"

using ComputeRectangleArea = hardware_interfaces::srv::ComputeRectangleArea;

class RectangleAreaServer : public rclcpp::Node
{
public:
  RectangleAreaServer() : Node("rectangle_area_server")
  {
    // Service 回调有 request、response 两个参数,std::bind 用 _1、_2 两个占位符对应
    service_ = create_service<ComputeRectangleArea>(
      "compute_rectangle_area",
      std::bind(&RectangleAreaServer::callbackArea, this, std::placeholders::_1, std::placeholders::_2));
    RCLCPP_INFO(get_logger(), "Rectangle area service ready on /compute_rectangle_area");
  }

private:
  // 读取 request 字段计算面积,写入 response;回调返回后框架自动把结果发回 Client
  void callbackArea(
    const ComputeRectangleArea::Request::SharedPtr request,
    const ComputeRectangleArea::Response::SharedPtr response)
  {
    response->area = request->length * request->width;
    RCLCPP_INFO(
      get_logger(), "%.2f x %.2f = %.2f",
      request->length, request->width, response->area);
  }

  rclcpp::Service<ComputeRectangleArea>::SharedPtr service_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<RectangleAreaServer>());
  rclcpp::shutdown();
  return 0;
}

File: extra-code/stage2_code/py_hardware_tutorial/py_hardware_tutorial/rectangle_area_server.py
# ============================================================
# 自定义服务接口示例：服务端（Server）
# 演示概念：自定义 srv（ComputeRectangleArea）的
# request/response 模型
# 提供 /compute_rectangle_area 服务：输入长和宽，返回面积
# 运行方式：
#   ros2 run py_hardware_tutorial rectangle_area_server
# ============================================================
import rclpy
# 自定义服务接口：Request 含 length、width，Response 含 area
from hardware_interfaces.srv import ComputeRectangleArea
from rclpy.node import Node


class RectangleAreaServer(Node):
    def __init__(self):
        super().__init__("rectangle_area_server")
        # 创建 Service：接口类型、服务名、处理回调
        self.service_ = self.create_service(
            ComputeRectangleArea, "compute_rectangle_area", self.callback_area)
        self.get_logger().info("Rectangle area service ready on /compute_rectangle_area")

    def callback_area(self, request, response):
        # 从 request 取输入，算好结果填进 response 再返回，
        # 框架负责把 response 发回客户端
        response.area = request.length * request.width
        self.get_logger().info(
            f"{request.length:.2f} x {request.width:.2f} = {response.area:.2f}")
        return response


def main(args=None):
    rclpy.init(args=args)
    node = RectangleAreaServer()
    # 服务端必须一直 spin 才能持续响应请求
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()
